pilot/pkg/xds/ads.go

```diff
@@ -1007,33 +994,38 @@ func (conn *Connection) Watched(typeUrl string) *model.WatchedResource {
 	return nil
 }
 
-// pushDetails returns the details needed for current push. It returns ordered list of
-// watched resources for the proxy, ordered in accordance with known push order.
-// It also returns the lis of typeUrls.
-// nolint
-func (conn *Connection) pushDetails() ([]*model.WatchedResource, sets.String) {
-	conn.proxy.RLock()
-	defer conn.proxy.RUnlock()
-	typeUrls := sets.New[string]()
-	for k := range conn.proxy.WatchedResources {
-		typeUrls.Insert(k)
+// orderedXdsPush pushes xds to the connection in order.
+func (s *DiscoveryServer) orderedXdsPush(
+	conn *Connection,
+	pushRequest *model.PushRequest,
+	push func(con *Connection, w *model.WatchedResource, req *model.PushRequest) error,
+) error {
+	// first push all known types, in order
+	for _, tp := range PushOrder {
+		conn.proxy.RLock()
```

Comment:
I think this has two issues
1. We are not ordering the entire PushOrder once as we used to do earlier. We are reading one typeURL at a time which may lead to inconsistencies. Not sure what impacts it will have
2. Locking/Releasing multiple times - not sure if it is slower than locking once and reading